Add database tables

rrennick opened the pull request:
Fixes #77 
Implements https://github.com/Prospress/action-scheduler-custom-tables/issues/48 .

As of https://github.com/Prospress/action-scheduler/commit/97337a8718607ab745efc8b830271df10c751088 the PR works with PHP 7.3.

Outstanding items 
- [x] ~~PHP 5.3 compatibility~~ Migration & custom tables are disabled in PHP versions <5.6. 
- [x] Unit tests
- [x] Unit test failures
- [x] Tweak/bugfix small inconsistencies (eg. the migration notice shows even when the migration is disabled by a custom data store).

### Tests Scenarios

- [ ] Upgrade site from Action Scheduler 2.2.n and confirm migration completes correctly
- [ ] Upgrade site from Action Scheduler Custom Tables plugin and confirm no errors or interference with previously migrated actions

rrennick commented:
The planned date for updating WP to PHP 5.6 minimum is April - https://make.wordpress.org/core/2018/12/08/updating-the-minimum-php-version/ . Rather than make the DB tables data store compatible with PHP 5.3, having a 5.6 dependency on enabling it would be better.

rrennick commented:
@JPry Have you encountered the error in https://travis-ci.org/Prospress/action-scheduler/jobs/507947421 ? These tests run successfully locally so I'm wondering if this is a phpunit version issue?

JPry commented:
@rrennick I haven't seen that particular error before. However, I can see that an older version of PHPUnit is being run: version `4.8.27`. I believe we should be using version 5+. This may require downloading the correct version of PHPUnit as part of the pre-test setup.

rrennick commented:
@JPry Thanks, that worked. The second issue is that the `phpunit -c tests/phpunit.xml.dist --exclude-group migration` does exclude the test in my local env that is failing at

https://travis-ci.org/Prospress/action-scheduler/jobs/508379185#L139

Could that be related to the version of the WP tests lib?

JPry commented:
@rrennick It looks like that particular error is triggered on PHP 5.3 when using the short array syntax. I also see that the group on that file is `tables`, not `migration`

rrennick commented:
>  I also see that the group on that file is tables, not migration

Both groups are excluded. I verified that the phpunit command in my previous comment does exclude that test in my local environment.

JPry commented:
Perhaps it still has to parse the file in order to determine if it's part of the group, and it's the parsing that is triggering the error. 

I think any files that require a certain PHP version need to be included in a separate test suite with a version dependency. This is described here: https://phpunit.readthedocs.io/en/8.0/configuration.html#test-suites

rrennick commented:
> Perhaps it still has to parse the file in order to determine if it's part of the group, and it's the parsing that is triggering the error.

That sounds like the ticket.

rrennick commented:
This one is ready for review. If we could get this into the 3.0.0 branch then we could look at merging in master, addressing conflicts with other PRs, get some testing going & fix any follow up issues.

rcstr commented on classes/ActionScheduler_wcSystemStatus.php:
```diff
@@ -0,0 +1,147 @@
+	 * is triggered when invoking inaccessible methods in an object context.
+	 *
+	 * @param $name      string
+	 * @param $arguments array
```
@params are in the wrong order, type goes first then the variable name
```suggestion
	 * @param array $argument
```

rrennick replied:
This file also not in the PR.

rcstr commented on classes/WP_CLI/ActionScheduler_WPCLI_QueueRunner.php:
```diff
@@ -177,6 +174,7 @@ public function action_failed( $action_id, $exception ) {
 	 * @param int $sleep_time Amount of seconds to sleep
+	 * @deprecated 3.0.0
 	 */
 	protected function stop_the_insanity( $sleep_time = 0 ) {
```
Missing `_deprecated_function` call

rrennick commented:
@rcstr Thanks for looking at this :) I made a copy of this branch at https://github.com/Prospress/action-scheduler/tree/add/custom-tables-old and recreated this branch with a clean commit history. 

I have a copy of your comments in email if any of them disappear from this PR. I'll ensure all are addressed either on this PR or a separate one if the affected code isn't in the clean commit history.

rrennick commented on docs/assets/css/style.scss:
```diff
@@ -29,4 +29,29 @@ footer {
+	animation:octocat-wave 560ms ease-in-out
+}
+
+@keyframes octocat-wave{
```
This file is no longer in the PR so I will address this one separately.

rrennick commented on docs/assets/css/style.scss:
```diff
@@ -29,4 +29,29 @@ footer {
+	.github-corner:hover .octo-arm {
+		animation:none
+	}
+	.github-corner .octo-arm{
```
Same as the previous & not in the PR.

rrennick commented on classes/schema/ActionScheduler_StoreSchema.php:
```diff
@@ -0,0 +1,81 @@
+				        KEY last_attempt_gmt (last_attempt_gmt),
+				        KEY last_attempt_local (last_attempt_local),
+				        KEY claim_id (claim_id)
+				        ) $charset_collate";
```
The GMT date fields are using to order/filter queries so those fields should have indexes. The args field is used for search in the dashboard. I don't see the `*_local` date fields in any WHERE/ORDER BY clauses.

@JPry Do you know of any users of the ASCT plugin using the `*_local` indexes?

JPry replied:
@rrennick I don't know of anything taking advantage of the indexes on `*_local` tables.

rrennick commented:
@rcstr I've got back to this and have finished the remaining review items.

rcstr commented:
Hi @rrennick - My apologies, didn't see the ping earlier, I'm AFK this week but scheduled this to review again on Monday next week

thenbrent commented on classes/ActionScheduler_Data.php:
```diff
@@ -0,0 +1,185 @@
+ * Class ActionScheduler_Data
+ *
+ * The main plugin/initialization class for the
+ * Action Scheduler Custom Tables plugin.
```
The class is no longer used in the Custom Tables plugin.

thenbrent commented on classes/ActionScheduler_Data.php:
```diff
@@ -0,0 +1,185 @@
+	 *
+	 * @return string
+	 */
+	public function get_store_class( $class ) {
```
We have the `'action_scheduler_store_class'` filter already here: https://github.com/Prospress/action-scheduler/blob/2.2.5/classes/ActionScheduler_Store.php#L196

That's what this method is being attached to. This confusion highlights that the logic in this method may be better moved to `ActionScheduler_Store::instance()` and this method can be deleted (it was required when the code was in a plugin, it's not required anymore).

rrennick replied:
This was moved to the data controller class.

thenbrent commented on classes/ActionScheduler_Data.php:
```diff
@@ -0,0 +1,185 @@
+	public function get_logger_class( $class ) {
+		ActionScheduler_Store::instance();
+
+		if ( $this->store_classname ) {
```
I think we could move this code to `ActionScheduler_Logger::instance()`. Do we need to default to `'ActionScheduler_wpCommentLogger'` at all in core anymore?

This method only seems needed when overriding the default store from a plugin.

rrennick replied:
> Do we need to default to 'ActionScheduler_wpCommentLogger' at all in core anymore?

It's needed for the migration.

rrennick replied:
Added function `has_custom_datastore()`

thenbrent commented on classes/ActionScheduler_Data.php:
```diff
@@ -0,0 +1,185 @@
+	public function register_cli_command() {
+		if ( defined( 'WP_CLI' ) && WP_CLI && $this->allow_custom_migration() && ! $this->migration_scheduler->is_migration_complete() && $this->migration_scheduler->dependencies_met() ) {
+			$command = new ActionScheduler_WPCLI_Migration_Command();
+			$command->register();
```
Could we unify the registration of this command with the registration of the `ActionScheduler_WPCLI_Scheduler_command` command rather than having two separate CLI systems.

rrennick replied:
> This confusion highlights that the logic in this method may be better moved to ActionScheduler_Store::instance() and this method can be deleted

AS is compatible with PHP 5.2 but the migration to CT won't kick in unless PHP is >= 5.6. I opted for adding filters in the CT code over adding 5.6 conditional checks in the main plugin code. That can be revisited. I'll look at better filter names to start.

rrennick replied:
> Could we unify the registration of this command

The reason I added it that way is the command should only be registered if migration conditions are met and the migration isn't complete. The idea being that once the migration is done, none of the migration code is loaded.

thenbrent replied:
> the migration to CT won't kick in unless PHP is >= 5.6

Ah, I missed that critical part. That makes sense, and is a really good approach. While WP 5.2 & WC 3.7 require PHP 5.6, I like the idea of maintaining backward compatibility with 5.2 for a while longer in AS for older users.

thenbrent commented on classes/ActionScheduler_Data.php:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @codeCoverageIgnore
+ */
+class ActionScheduler_Data {
```
There's a few methods in this class which would be better suited elsewhere (I've left comments on them individually). Once they're moved, the class seems to be focused primarily on migration, so it may be better off renamed to something conveying that.

thenbrent commented on classes/WP_CLI/ActionScheduler_WPCLI_Migration_Command.php:
```diff
@@ -0,0 +1,130 @@
+	 * Register the command with WP-CLI
+	 */
+	public function register() {
+		if ( ! defined( 'WP_CLI' ) || ! WP_CLI ) {
```
Can `WP_CLI` be disabled in certain contexts where `WP_CLI_Command` may exist? It may be an oversight, but I have a lot of faith in the original author of this, so want to be sure he didn't know something we don't.

Regardless, we may want to move this to here: https://github.com/Prospress/action-scheduler/blob/2.2.5/classes/ActionScheduler.php#L112

That would avoid having two different registration approaches for the two different WP CLI commands.

thenbrent commented on classes/WP_CLI/ActionScheduler_WPCLI_Migration_Command.php:
```diff
@@ -0,0 +1,130 @@
+			return;
+		}
+
+		WP_CLI::add_command( 'action-scheduler db-tables migrate', [ $this, 'migrate' ], [
```
The `db-tables` component of this command feels unnecessary now that the code is in AS core. What about just `action-scheduler migrate`? (Or a similar keyword used by WC if it differs)

thenbrent commented on classes/WP_CLI/ActionScheduler_WPCLI_QueueRunner.php:
```diff
@@ -112,7 +114,6 @@ public function run() {
 
 			$this->process_action( $action_id );
 			$this->progress_bar->tick();
-			$this->maybe_stop_the_insanity();
```
It looks like `ActionScheduler_WPCLI_ProgressBar::free_memory` is being called by `ActionScheduler_WPCLI_ProgressBar::tick` now.

It's surprising that a `ProgressBar` is also responsible for memory management though.

rrennick replied:
`free_memory()` was moved to the data controller class.

thenbrent commented:
I've just added a couple of Tests Scenarios to the main description on this PR.

rrennick commented:
> What do you think? Do you think we can rename some of the classes and remove the prefix?

@rcstr Done in https://github.com/Prospress/action-scheduler/pull/259/commits/78cf42788c1f7867e3e2a1f4cd5dbd5795104c90

rrennick commented:
@rcstr Updated PHPDocs in https://github.com/Prospress/action-scheduler/pull/259/commits/5370f4bb8ea3afec66cbaceee374f6732bda248b

rrennick commented on classes/data-stores/ActionScheduler_DBStore.php:
```diff
@@ -0,0 +1,625 @@
+	 *
+	 * @return void
+	 */
+	public function log_execution( $action_id ) {
```
The reason for the query instead of the update function is `attempts = attempts+1`. If you retrieve the record, increment & save it's possible to overwrite an update that occurred while in process.

rrennick commented:
@rcstr @thenbrent Thanks so much for the recommendations. I've updated the PR with

- rename WP CLI command to `migrate`
- move CLI command to common register block
- split ActionScheduler_Data into data controller and migration controller
- move dependencies & migration complete to data controller
- no migration related code is loaded unless the site is eligible for migration
- rename migration hook group
- removed `custom_` from a few hooks, strings, etc.
- fix help on CLI run command
- move free_memory to data controller
- use free_memory() in stop_the_insanity()
- add `free-memory-on` and `pause` parameters to CLI commands
- check for object cache object
- other suggested fixes
- unit test class name usage changes

The failed unit test is the one fixed in #300. Some of the unit tests should probably be moved to a new file for the data controller but that can be done in a follow up PR.

rcstr commented:
Thank you @rrennick - I will look into the changes early next week 👍

rrennick commented:
> do you have a recommendation for what would be the best way to test?

What I do is make a backup of my DB so I can restore to re-run the migration as many times as needed. Then 

- use a snippet of code like https://github.com/Prospress/action-scheduler/pull/323 (for that snippet you need to switch to that branch) to generate a few thousand actions.
- make a second backup so you can re-run migrations
- migrate with WP CLI
- let another one process on the normal cron
- I'd be interested in seeing the time comparison of the migration on the async PR as well

rrennick commented on classes/data-stores/ActionScheduler_DBStore.php:
```diff
@@ -0,0 +1,715 @@
+	public function release_claim( ActionScheduler_ActionClaim $claim ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+		$wpdb->update( $wpdb->actionscheduler_actions, [ 'claim_id' => 0 ], [ 'claim_id' => $claim->get_id() ], [ '%d' ], [ '%d' ] );
```
I think in this case we want to remove the claim record whether or not removing the claim from the actions is successful. 

The number of active batches is determined by counting the claim records: https://github.com/Prospress/action-scheduler/blob/add/custom-tables/classes/data-stores/ActionScheduler_DBStore.php#L556:L563 Leaving these behind could eventually disable starting new batches.

thenbrent replied:
@rrennick thinking is right above. I'll resolve this.

rrennick commented on classes/data-stores/ActionScheduler_DBStore.php:
```diff
@@ -0,0 +1,715 @@
+<?php
```
fixed in https://github.com/Prospress/action-scheduler/pull/259/commits/917b613000d897292a200ab50f1fb01b972b622c

rrennick commented:
@thenbrent @rcstr This has had the merge conflicts fixed & unit tests updated to work with merged changes.

thenbrent commented:
Merging this one. Nick work @rrennick! Thanks for the reviews @rcstr! 🎉 

I gave all files a final review today, comparing all files to their original in the Custom Tables plugin and reviewing carefully all newly introduced code. I have a number of outstanding items I'd like addressed, but there was only one bug, the rest are just refactoring or tweaks, which we can submit in separate PRs rather than blocking this one (which has merge conflicts with other open PRs and is generally an enormous diff to keep coming back to for small changes).

> I still have a few files left to review, then I will proceed with QA

@rcstr given the large number of other PRs open for 3.0.0, instead of spending QA time on this, I recommend from here, we focus on QA of the first beta, as it will help test other important changes too, like the async runner, and potentially refactor of recurring schedules #333.
